Air shuffles should not use fringe destination as scratch

justinmichaud opened the pull request:
#### 2cba369d0b95a5e3a07452103efbb20d16ca3de0
<pre>
Air shuffles should not use fringe destination as scratch
<a href="https://bugs.webkit.org/show_bug.cgi?id=283641">https://bugs.webkit.org/show_bug.cgi?id=283641</a>

Reviewed by Yusuke Suzuki.

When we have a shuffle in air, we have a series of shifts
and rotates. Rotates have extra bits not part of the loop,
called the fringe.

A series of shifts permits the destination of the first shift
to be used as scratch before it gets its final value. This is
not true for a fringe, even though a fringe is basically a shift.
This is only because the lifetime of the scratch register extends
into the rotate that follows it.

Basically, the mutable state in this code gets very confused
because of the constantly changing order of instructions emitted.

This bug was discovered with ARMv7 wasm tail calls, and the test case
will be included in that patch.

* Source/JavaScriptCore/b3/air/AirEmitShuffle.cpp:
(JSC::B3::Air::emitShuffle):

Canonical link: <a href="https://commits.webkit.org/287139@main">https://commits.webkit.org/287139@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2eca7e262ca9336a2680e0803be1b709de28cd02

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78470 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57515 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31852 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83131 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29735 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66666 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61455 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19373 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81537 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51469 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69343 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41767 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28072 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71616 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25600 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84497 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77707 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5835 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3977 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69681 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5996 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/67455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68936 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17184 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12956 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11343 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100016 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5782 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21840 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5770 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9204 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/air/AirEmitShuffle.cpp:
```diff
@@ -359,6 +359,8 @@ Vector<Inst> emitShuffle(
         result.append(Inst(move, origin, pair.src(), pair.dst()));
     };
 
+    bool freezeScratches = false;
```
Instead of having a bool variable, can we pass a bool variable to `handleShift` to make it explicit from the caller?

Constellation reviewed (approved):
r=me. Is it possible to add a test for testb3 / testair?

justinmichaud committed:
Air shuffles should not use fringe destination as scratch
https://bugs.webkit.org/show_bug.cgi?id=283641

Reviewed by Yusuke Suzuki.

When we have a shuffle in air, we have a series of shifts
and rotates. Rotates have extra bits not part of the loop,
called the fringe.

A series of shifts permits the destination of the first shift
to be used as scratch before it gets its final value. This is
not true for a fringe, even though a fringe is basically a shift.
This is only because the lifetime of the scratch register extends
into the rotate that follows it.

Basically, the mutable state in this code gets very confused
because of the constantly changing order of instructions emitted.

This bug was discovered with ARMv7 wasm tail calls, and the test case
will be included in that patch.

* Source/JavaScriptCore/b3/air/AirEmitShuffle.cpp:
(JSC::B3::Air::emitShuffle):

Canonical link: https://commits.webkit.org/287139@main
